src/input/mod.rs

```diff
@@ -305,6 +307,32 @@ impl State {
         self.handle_bind(bind);
     }
 
+    fn apply_keyboard_config<I: InputBackend>(&mut self, event: &I::KeyboardKeyEvent) {
+        let device = event.device();
+
+        let keyboard_config = self
+            .niri
+            .config
+            .borrow()
+            .input
+            .keyboard_named(device.name());
+
+        if keyboard_config != self.niri.current_keyboard {
+            let keyboard = self.niri.seat.get_keyboard().unwrap();
+
+            if let Err(err) = keyboard.set_xkb_config(self, keyboard_config.xkb.to_xkb_config()) {
```

Comment:
Would it be worth to check the xkb config and the repeat info for equality individually too?

Reply:
it now checks the key repeat settings and xkb independently and only applies what needs changing.